src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDependencyDao.java

```diff
@@ -65,36 +73,28 @@ public boolean datasetHasSnapshotReference(Dataset dataset) {
         return hasDependencies;
     }
 
-    private boolean hasReference(Query query) throws InterruptedException {
-        ApiFuture<QuerySnapshot> querySnapshot = query.get();
-
-        try {
-            List<QueryDocumentSnapshot> documents = querySnapshot.get().getDocuments();
-            return (documents.size() > 0);
-        } catch (ExecutionException ex) {
-            throw new FileSystemExecutionException("has reference - execution exception", ex);
-        }
-    }
-
     public List<String> getDatasetSnapshotFileIds(Dataset dataset, String snapshotId) throws InterruptedException {
         DatasetDataProject dataProject = dataLocationService.getProjectOrThrow(dataset);
         FireStoreProject fireStoreProject = FireStoreProject.get(dataProject.getGoogleProjectId());
         String dependencyCollectionName = getDatasetDependencyId(dataset.getId().toString());
         CollectionReference depColl = fireStoreProject.getFirestore().collection(dependencyCollectionName);
+
         Query query = depColl.whereEqualTo("snapshotId", snapshotId);
-        ApiFuture<QuerySnapshot> querySnapshot = query.get();
+        int batchSize = configurationService.getParameterValue(FIRESTORE_QUERY_BATCH_SIZE);
+        FireStoreBatchQueryIterator queryIterator = new FireStoreBatchQueryIterator(query, batchSize);
```

Comment:
Instead of implementing a pseudo-iterator, can you make `FireStoreBatchQueryIterator implement Iterable`? I think this could simplify the for loop below and make things a bit more clear.

Reply:
I don't see the benefit. It would have made the iterator implementation much more complex. To make it implement Iterable, It would have to take a lambda. The callers could not use standard Consumers, because those don't allow checked exceptions. So to me it seems like a lot of work for minor aesthetics.